block-storage/src/main/scala/coop/rchain/blockstorage/dag/DagMessageState.scala

```diff
@@ -0,0 +1,188 @@
+package coop.rchain.blockstorage.dag
+
+import cats.syntax.all._
+
+// DagMessageState represents state of one validator in the network
+final case class DagMessageState[M: Ordering, S: Ordering](
+    me: S,
+    latestMsgs: Set[Message[M, S]],
+    // Message view - updated when new message is added
+    msgViewMap: Map[M, Message[M, S]] = Map()
+) {
+
+  /**
+    * Creates a new message, generates id (hash) and finalization fringe
+    */
+  def createMessageView(
+      id: M,
+      height: Long,
+      sender: S,
+      senderSeq: Long,
+      finBondsMap: Map[S, Long],
+      justifications: Set[Message[M, S]]
+  ): Message[M, S] = {
+    // Calculate next fringe or continue with parent
+    val finalizer                    = Finalizer(msgViewMap)
+    val (parentFringe, newFringeOpt) = finalizer.calculateFinalization(justifications, finBondsMap)
+
+    val newFringe    = newFringeOpt.getOrElse(parentFringe)
+    val newFringeIds = newFringe.map(_.id)
+
+    // Seen messages are all seen from justifications combined
+    val seenByParents = justifications.flatMap(_.seen)
+    val newSeen       = seenByParents + id
+
+    // Create message view, an immutable object with all fields calculated
+    val justificationKeys = justifications.map(_.id)
+    val newMsgView =
+      Message(
+        id,
+        height,
+        sender,
+        senderSeq,
+        finBondsMap,
+        justificationKeys,
+        fringe = newFringeIds,
+        seen = newSeen
+      )
+
+    /* Debug log */
+    if (me == sender) {
+      //        val fringeStr = newFringeOpt.map(ms => s"+ ${showMsgs(ms)}").getOrElse("")
+      //        println(s"${newMsgView.id} $fringeStr")
+      debugLogMessage(finalizer, newMsgView, justifications, parentFringe, newFringeOpt)
+    }
+    /* End Debug log */
+
+    newMsgView
+  }
+
+  /**
+    * Inserts a message to sender's state
+    */
+  def insertMsgView(msgView: Message[M, S]): DagMessageState[M, S] =
```

Comment:
Might be it would be better to name methods as `createMsg`, `insertMsg`, like `createMsgAndUpdateSender` below. As you work with `Message` case class.

Reply:
I renamed everywhere to just use a message instead of message view.